File: built-in plugins dsp/mb-mallet-xylophone.cpp
/**
 * MB Xylophone
 * Category : instrument
 * Type     : mallet
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Bright wooden xylophone with sharp attack
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MALLET_XYLOPHONE_H
#define MB_MALLET_XYLOPHONE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMalletXylophone : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mallet-xylophone";
    static constexpr const char* PLUGIN_NAME    = "MB Xylophone";
    static constexpr const char* PLUGIN_TYPE    = "mallet";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float brightness = 0.8f;  // range [0, 1]
    float mallet = 0.7f;  // range [0, 1]
    float resonance = 0.3f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbMalletXylophone() = default;
    ~MbMalletXylophone() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.brightness = std::clamp(params.brightness, 0f, 1f);
        params.mallet = std::clamp(params.mallet, 0f, 1f);
        params.resonance = std::clamp(params.resonance, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Xylophone
        return input;
    }
};

#endif // MB_MALLET_XYLOPHONE_H

File: built-in plugins dsp/mb-mallet-xylophone.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbMalletXylophonePlugin: PluginDefinition = {
  id: "mb-mallet-xylophone",
  slug: "mb-mallet-xylophone",
  name: "MB Xylophone",
  category: "instrument",
  type: "mallet" as any,
  version: "1.0.0",
  description: "Bright wooden xylophone with sharp attack",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sine", detune: 0, gain: 0.5 },
    { type: "sine", detune: 2400, gain: 0.3 },
    { type: "triangle", detune: 3600, gain: 0.2 },
  ],
  envelope: { attack: 0.001, decay: 0.4, sustain: 0.0, release: 0.3 },
  parameters: [
    {
      id: "brightness",
      name: "Brightness",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "mallet",
      name: "Mallet Hardness",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "resonance",
      name: "Resonance",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { brightness: 0.8, mallet: 0.7, resonance: 0.3, volume: 0.8 },
};

export default MbMalletXylophonePlugin;
